src/main.rs

```diff
@@ -139,204 +174,209 @@ impl DiemDemo {
-              "recipient_address": "0x57c76da2e144c0357336ace2f3f8ac9b",
-              "amount": 10,
-              "metadata_hex_encoded": "",
-              "metadata_signature_hex_encoded": ""
-            }
-          }
-        });
-        let mut g_res: GenerateRawTxnRequest = serde_json::from_value(tx_json).unwrap();
-        g_res.txn_params.sender_address = String::from("0x").add(self.account.as_ref().unwrap().address.to_string().as_str());
-        let g_resp:GenerateRawTxnResponse = generate_raw_txn(g_res);
-        let s_res: SignTransactionUsingEd25519Request = SignTransactionUsingEd25519Request{
-            raw_txn: g_resp.raw_txn.clone(),
-            private_key: self.account.as_ref().unwrap().key_pair.as_ref().unwrap().private_key.to_encoded_string().unwrap().clone(),
-        };
-        let s_resp:SignTransactionUsingEd25519Response = sign_transaction_using_ed25519(s_res);
-        let gs_res:GenerateSignedTxnRequest = GenerateSignedTxnRequest{
-            raw_txn: g_resp.raw_txn.clone(),
-            public_key: self.account.as_ref().unwrap().key_pair.as_ref().unwrap().public_key.to_encoded_string().unwrap().clone(),
-            signature: s_resp.signature.clone(),
-        };
-        let gs_resp:GenerateSignedTxnResponse = generate_signed_txn(gs_res);
-
-        println!("{:#?}", gs_resp.signed_txn);
-        Ok(())
-    }
-
-    pub fn init_state(
-        &mut self
-    ) -> Result<()> {
+    async fn init_state(
+        &mut self,
+        pr: Option<&PrClient>,
+    ) -> Result<(), Error> {
         let mut batch = JsonRpcBatch::new();
         batch.add_get_state_proof_request(0);
+        if let Ok(resp) = self.request_rpc(batch) {
+            let state_proof = StateProofView::from_response(resp).unwrap();
+            //println!("state_proof:\n{:?}", state_proof);
+
+            let epoch_change_proof: EpochChangeProof =
+                bcs::from_bytes(&state_proof.epoch_change_proof.into_bytes().unwrap()).unwrap();
+            let ledger_info_with_signatures: LedgerInfoWithSignatures =
+                bcs::from_bytes(&state_proof.ledger_info_with_signatures.into_bytes().unwrap()).unwrap();
+
+            // Init zero version state
+            let zero_ledger_info_with_sigs = epoch_change_proof.ledger_info_with_sigs[0].clone();
+
+            self.latest_epoch_change_li = Option::from(zero_ledger_info_with_sigs.clone());
+            self.trusted_state = Option::from(TrustedState::try_from(zero_ledger_info_with_sigs.ledger_info()).unwrap());
+            self.latest_li = Option::from(ledger_info_with_signatures.clone());
+            self.epoch_change_proof = Option::from(epoch_change_proof.clone());
+
+            // Update Latest version state
+            let _ = self.verify_state_proof(ledger_info_with_signatures, epoch_change_proof);
+            println!("trusted_state: {:#?}", self.trusted_state);
+            println!("ledger_info_with_signatures: {:#?}", self.latest_li);
```

Comment:
I think `verify_state_proof` is already inside the TEE. We can feel free to remove this additional verification code here.

Now the only place we really use `trusted_state` is `get_transaction_proof()`. We get the latest ledger version from `trusted_state`. Is it possible to get the version from `state_proof` directly?